File: wp_to_thelia_converter.php
<?php
/**
 * Script pour convertir les données WordPress/WooCommerce vers des CSV compatibles Thelia
 */

class WordPressToTheliaConverter {
    private $sqlFile;
    private $posts = [];
    private $terms = [];
    private $termTaxonomy = [];
    private $postmeta = [];
    private $termRelationships = [];
    
    public function __construct($sqlFile) {
        $this->sqlFile = $sqlFile;
    }
    
    /**
     * Extrait les données d'une table SQL
     */
    private function parseSqlTable($tableName) {
        echo "Extraction de la table: $tableName\n";
        
        if (!file_exists($this->sqlFile)) {
            echo "Fichier SQL non trouvé: " . $this->sqlFile . "\n";
            return [];
        }
        
        $content = file_get_contents($this->sqlFile);
        
        // Pattern pour trouver les INSERT dans une table spécifique
        $pattern = "/INSERT INTO `$tableName`.*?VALUES\s*(.+?);/si";
        
        if (!preg_match($pattern, $content, $matches)) {
            echo "Aucune donnée trouvée pour la table $tableName\n";
            return [];
        }
        
        $valuesStr = $matches[1];
        $rows = [];
        
        // Parser les valeurs multi-lignes
        $currentRow = '';
        $parenthesesCount = 0;
        $inQuotes = false;
        $i = 0;
        $len = strlen($valuesStr);
        
        while ($i < $len) {
            $char = $valuesStr[$i];
            
            if ($char === "'" && ($i === 0 || $valuesStr[$i - 1] !== '\\')) {
                $inQuotes = !$inQuotes;
                $currentRow .= $char;
                $i++;
            } elseif (!$inQuotes && $char === '(') {
                $parenthesesCount++;
                $currentRow .= $char;
                $i++;
            } elseif (!$inQuotes && $char === ')') {
                $parenthesesCount--;
                $currentRow .= $char;
                $i++;
                
                if ($parenthesesCount === 0) {
                    // Fin d'une ligne
                    if (preg_match('/^\((.*)\)$/', trim($currentRow), $rowMatch)) {
                        $row = $this->parseInsertValues($rowMatch[1]);
                        if (!empty($row)) {
                            $rows[] = $row;
                        }
                    }
                    $currentRow = '';
                    
                    // Sauter les virgules et espaces
                    while ($i < $len && ($valuesStr[$i] === ',' || $valuesStr[$i] === ' ' || $valuesStr[$i] === "\n")) {
                        $i++;
                    }
                }
            } else {
                $currentRow .= $char;
                $i++;
            }
        }
        
        echo "Table $tableName: " . count($rows) . " enregistrements trouvés\n";
        return $rows;
    }
    
    /**
     * Parse une ligne de valeurs INSERT SQL
     */
    private function parseInsertValues($valuesStr) {
        $values = [];
        $current = '';
        $inQuotes = false;
        $i = 0;
        $len = strlen($valuesStr);
        
        while ($i < $len) {
            $char = $valuesStr[$i];
            
            if ($char === '\\' && $i + 1 < $len) {
                // Caractère échappé
                $current .= $valuesStr[$i + 1];
                $i += 2;
            } elseif ($char === "'" && ($i === 0 || $valuesStr[$i - 1] !== '\\')) {
                // Début/fin de chaîne
                $inQuotes = !$inQuotes;
                $current .= $char;
                $i++;
            } elseif ($char === ',' && !$inQuotes) {
                // Séparateur de valeurs
                $values[] = $this->cleanValue($current);
                $current = '';
                $i++;
            } else {
                $current .= $char;
                $i++;
            }
        }
        
        // Ajouter la dernière valeur
        if (!empty($current)) {
            $values[] = $this->cleanValue($current);
        }
        
        return $values;
    }
    
    /**
     * Nettoie une valeur extraite
     */
    private function cleanValue($value) {
        $value = trim($value);
        
        if ($value === 'NULL') {
            return '';
        }
        
        if (preg_match("/^'(.*)'$/", $value, $matches)) {
            $value = $matches[1];
            // Gérer les échappements
            $value = str_replace("\\'", "'", $value);
            $value = str_replace("\\\"", "\"", $value);
            $value = str_replace("\\\\", "\\", $value);
        }
        
        return $value;
    }
    
    /**
     * Extrait toutes les données des tables WordPress
     */
    public function extractAllData() {
        echo "Début de l'extraction des données WordPress...\n";
        
        $this->posts = $this->parseSqlTable('wp_posts');
        $this->terms = $this->parseSqlTable('wp_terms');
        $this->termTaxonomy = $this->parseSqlTable('wp_term_taxonomy');
        $this->postmeta = $this->parseSqlTable('wp_postmeta');
        $this->termRelationships = $this->parseSqlTable('wp_term_relationships');
        
        echo "Extraction terminée!\n";
    }
    
    /**
     * Récupère les metas d'un post
     */
    private function getPostMetas($postId) {
        $metas = array();
        foreach ($this->postmeta as $meta) {
            if (isset($meta[1]) && $meta[1] == $postId) {
                $key = isset($meta[2]) ? $meta[2] : '';
                $value = isset($meta[3]) ? $meta[3] : '';
                $metas[$key] = $value;
            }
        }
        return $metas;
    }
    
    /**
     * Récupère le nom d'un terme
     */
    private function getTermName($termId) {
        foreach ($this->terms as $term) {
            if (isset($term[0]) && $term[0] == $termId) {
                return isset($term[1]) ? $term[1] : '';
            }
        }
        return '';
    }
    
    /**
     * Récupère les catégories d'un produit
     */
    private function getProductCategories($postId) {
        $categories = array();
        
        foreach ($this->termRelationships as $rel) {
            if (isset($rel[0]) && $rel[0] == $postId) {
                $termTaxonomyId = isset($rel[1]) ? $rel[1] : '';
                
                foreach ($this->termTaxonomy as $tt) {
                    if (isset($tt[0]) && $tt[0] == $termTaxonomyId && (isset($tt[3]) ? $tt[3] : '') === 'category') {
                        $termId = isset($tt[1]) ? $tt[1] : '';
                        $categoryName = $this->getTermName($termId);
                        if ($categoryName) {
                            $categories[] = $categoryName;
                        }
                    }
                }
            }
        }
        
        return $categories;
    }
    
    /**
     * Génère le CSV des catégories
     */
    public function generateCategoriesCsv() {
        echo "Génération du CSV des catégories...\n";
        
        $categories = array();
        $processedTerms = array();
        
        foreach ($this->termTaxonomy as $tt) {
            if (isset($tt[3]) && $tt[3] === 'category') {
                $termId = isset($tt[1]) ? $tt[1] : '';
                
                if (!in_array($termId, $processedTerms)) {
                    $processedTerms[] = $termId;
                    
                    $categoryName = $this->getTermName($termId);
                    $description = '';
                    
                    // Récupérer la description
                    foreach ($this->terms as $term) {
                        if (isset($term[0]) && $term[0] == $termId) {
                            $description = isset($term[3]) ? $term[3] : '';
                            break;
                        }
                    }
                    
                    $categories[] = array(
                        'CATEGORIES FR' => $categoryName,
                        'CATEGORIES UK' => $categoryName,
                        'CHAPO FR' => mb_substr($description, 0, 200, 'UTF-8'),
                        'CHAPO UK' => mb_substr($description, 0, 200, 'UTF-8'),
                        'DESCRIPTIF FR' => $description,
                        'DESCRIPTIF UK' => $description,
                        'PHOTO' => ''
                    );
                }
            }
        }
        
        $this->writeCsv('setup/csvfromsql/categories.csv', $categories);
        echo "Categories CSV généré: " . count($categories) . " catégories\n";
    }
    
    /**
     * Génère le CSV des produits
     */
    public function generateProductsCsv() {
        echo "Génération du CSV des produits...\n";
        
        $products = array();
        
        foreach ($this->posts as $post) {
            if (isset($post[12]) && $post[12] === 'product') {
                $postId = isset($post[0]) ? $post[0] : '';
                $postTitle = isset($post[5]) ? $post[5] : '';
                $postContent = isset($post[11]) ? $post[11] : '';
                $postExcerpt = isset($post[9]) ? $post[9] : '';
                
                $metas = $this->getPostMetas($postId);
                
                // Informations de base
                $sku = array_key_exists('_sku', $metas) ? $metas['_sku'] : 'WP_' . $postId;
                $price = array_key_exists('_price', $metas) ? $metas['_price'] : '0';
                $regularPrice = array_key_exists('_regular_price', $metas) ? $metas['_regular_price'] : $price;
                $salePrice = array_key_exists('_sale_price', $metas) ? $metas['_sale_price'] : '';
                
                // Catégories
                $categories = $this->getProductCategories($postId);
                $categoryPath = !empty($categories) ? implode(' > ', $categories) : 'NON CLASSÉ';
                
                // Marque (attribut)
                $brand = '';
                foreach ($metas as $key => $value) {
                    if (strpos(strtolower($key), 'attribute') !== false && 
                        (strpos(strtolower($key), 'brand') !== false || strpos(strtolower($key), 'marque') !== false)) {
                        $brand = $value;
                        break;
                    }
                }
                
                // Image
                $image = '';
                $thumbnailId = array_key_exists('_thumbnail_id', $metas) ? $metas['_thumbnail_id'] : '';
                if ($thumbnailId) {
                    $thumbnailMetas = $this->getPostMetas($thumbnailId);
                    $imageFile = array_key_exists('_wp_attached_file', $thumbnailMetas) ? $thumbnailMetas['_wp_attached_file'] : '';
                    $image = basename($imageFile);
                }
                
                $products[] = array(
                    'REF' => $sku,
                    'TITRE UK' => $postTitle,
                    'CHAPO UK' => mb_substr($postExcerpt, 0, 200, 'UTF-8'),
                    'CHAPO FR' => mb_substr($postExcerpt, 0, 200, 'UTF-8'),
                    'DESCRIPTIF UK' => $postContent,
                    'DESCRIPTIF FR' => $postContent,
                    'POSTSCRIPTUM UK' => '',
                    'POSTSCRIPTUM FR' => '',
                    'PRIX' => $regularPrice,
                    'PRIX2' => $salePrice,
                    'PHOTO' => $image,
                    'BRAND' => $brand,
                    'COULEUR UK' => '',
                    'MATERIAL UK' => '',
                    'CONTENT UK' => '',
                    'CATEGORIE' => $categoryPath
                );
            }
        }
        
        $this->writeCsv('setup/csvfromsql/products.csv', $products);
        echo "Products CSV généré: " . count($products) . " produits\n";
    }
    
    /**
     * Génère le CSV des marques
     */
    public function generateBrandsCsv() {
        echo "Génération du CSV des marques...\n";
        
        $brands = array();
        
        foreach ($this->posts as $post) {
            if (isset($post[12]) && $post[12] === 'product') {
                $postId = isset($post[0]) ? $post[0] : '';
                $metas = $this->getPostMetas($postId);
                
                foreach ($metas as $key => $value) {
                    if (strpos(strtolower($key), 'attribute') !== false && 
                        (strpos(strtolower($key), 'brand') !== false || strpos(strtolower($key), 'marque') !== false)) {
                        if (!empty($value) && !in_array($value, $brands)) {
                            $brands[] = $value;
                        }
                    }
                }
            }
        }
        
        sort($brands);
        $brandsData = array();
        foreach ($brands as $brand) {
            $brandsData[] = array('brand' => $brand);
        }
        
        $this->writeCsv('setup/csvfromsql/brands.csv', $brandsData);
        echo "Brands CSV généré: " . count($brandsData) . " marques\n";
    }
    
    /**
     * Génère le CSV des couleurs
     */
    public function generateColorsCsv() {
        echo "Génération du CSV des couleurs...\n";
        
        $colors = array();
        
        foreach ($this->posts as $post) {
            if (isset($post[12]) && $post[12] === 'product') {
                $postId = isset($post[0]) ? $post[0] : '';
                $metas = $this->getPostMetas($postId);
                
                foreach ($metas as $key => $value) {
                    if (strpos(strtolower($key), 'attribute') !== false && 
                        (strpos(strtolower($key), 'color') !== false || strpos(strtolower($key), 'couleur') !== false)) {
                        if (!empty($value) && !in_array($value, $colors)) {
                            $colors[] = $value;
                        }
                    }
                }
            }
        }
        
        sort($colors);
        $colorsData = array();
        foreach ($colors as $color) {
            $colorsData[] = array('color' => $color);
        }
        
        $this->writeCsv('setup/csvfromsql/colors.csv', $colorsData);
        echo "Colors CSV généré: " . count($colorsData) . " couleurs\n";
    }
    
    /**
     * Génère le CSV des matériaux
     */
    public function generateMaterialsCsv() {
        echo "Génération du CSV des matériaux...\n";
        
        $materials = array();
        
        foreach ($this->posts as $post) {
            if (isset($post[12]) && $post[12] === 'product') {
                $postId = isset($post[0]) ? $post[0] : '';
                $metas = $this->getPostMetas($postId);
                
                foreach ($metas as $key => $value) {
                    if (strpos(strtolower($key), 'attribute') !== false && 
                        (strpos(strtolower($key), 'material') !== false || 
                         strpos(strtolower($key), 'matiere') !== false || 
                         strpos(strtolower($key), 'matériau') !== false)) {
                        if (!empty($value) && !in_array($value, $materials)) {
                            $materials[] = $value;
                        }
                    }
                }
            }
        }
        
        sort($materials);
        $materialsData = array();
        foreach ($materials as $material) {
            $materialsData[] = array('material' => $material);
        }
        
        $this->writeCsv('setup/csvfromsql/materials.csv', $materialsData);
        echo "Materials CSV généré: " . count($materialsData) . " matériaux\n";
    }
    
    /**
     * Écrit un fichier CSV
     */
    private function writeCsv($filename, $data) {
        if (empty($data)) {
            echo "Aucune donnée à écrire pour $filename\n";
            return;
        }
        
        $file = fopen($filename, 'w');
        if (!$file) {
            echo "Impossible d'ouvrir le fichier $filename\n";
            return;
        }
        
        // Entête
        $headers = array_keys($data[0]);
        fputcsv($file, $headers, ';');
        
        // Données
        foreach ($data as $row) {
            fputcsv($file, $row, ';');
        }
        
        fclose($file);
    }
    
    /**
     * Lance la conversion complète
     */
    public function convertAll() {
        echo "Début de la conversion WordPress vers Thelia...\n";
        echo "Fichier SQL: " . $this->sqlFile . "\n\n";
        
        // Extraire toutes les données
        $this->extractAllData();
        
        echo "\nGénération des fichiers CSV...\n";
        
        // Générer tous les fichiers CSV
        $this->generateCategoriesCsv();
        $this->generateProductsCsv();
        $this->generateBrandsCsv();
        $this->generateColorsCsv();
        $this->generateMaterialsCsv();
        
        echo "\nConversion terminée! Fichiers CSV générés dans setup/csvfromsql/\n";
    }
}

// Exécution du script
$converter = new WordPressToTheliaConverter('setup/uvpxgnzzfh.sql');
$converter->convertAll();
?>

File: wp_to_thelia_converter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Script pour extraire les données de WordPress/WooCommerce vers des CSV compatibles Thelia
"""

import re
import csv
import json
import os
from collections import defaultdict

class WordPressToTheliaConverter:
    def __init__(self, sql_file_path):
        self.sql_file_path = sql_file_path
        self.posts_data = []
        self.terms_data = []
        self.term_taxonomy_data = []
        self.postmeta_data = []
        self.term_relationships_data = []
        
        # Créer le répertoire de sortie
        self.output_dir = 'setup/csvfromsql'
        os.makedirs(self.output_dir, exist_ok=True)
        
    def parse_sql_insert(self, table_name):
        """Extrait les données d'INSERT SQL pour une table spécifique"""
        try:
            with open(self.sql_file_path, 'r', encoding='utf-8', errors='ignore') as file:
                content = file.read()
                
            # Find all INSERT statements for the table by looking for pattern more carefully
            # We'll look for INSERT INTO `table_name` ... VALUES and then find the ending );
            insert_pattern = rf"INSERT INTO `{table_name}`.*?VALUES\s*"
            insert_matches = list(re.finditer(insert_pattern, content, re.IGNORECASE))
            
            print(f"DEBUG: Found {len(insert_matches)} INSERT statements for table {table_name}")
            
            rows = []
            for i, match in enumerate(insert_matches):
                start_pos = match.end()
                
                # Find the end of this INSERT statement (look for ); at the end)
                remaining_content = content[start_pos:]
                
                # Look for the closing ); pattern that ends the INSERT
                end_match = re.search(r'\);', remaining_content)
                if end_match:
                    end_pos = start_pos + end_match.end()
                    insert_content = content[start_pos:end_pos]
                    
                    # Debug: Show we found an INSERT
                    # print(f"DEBUG: INSERT {i+1} processed, content length: {len(insert_content)}")
                    # print(f"DEBUG: First 100 chars of content: {insert_content[:100]}")
                    
                    # The content is already the VALUES part, so use it directly
                    values_str = insert_content
                    
                    # Replace newlines with spaces to make parsing easier
                    values_str = values_str.replace('\n', ' ').replace('\r', ' ')
                    
                    # print(f"DEBUG: VALUES string length: {len(values_str)}")
                    
                    # Use simple regex to split value groups
                    value_groups = re.findall(r'\([^)]*\)', values_str)
                    
                    # Debug: Show how many value groups we found
                    # if len(value_groups) > 0:
                    #     print(f"DEBUG: INSERT {i+1} found {len(value_groups)} value groups")
                    # else:
                    #     print(f"DEBUG: INSERT {i+1} found 0 value groups, values_str length: {len(values_str)}")
                    #     print(f"DEBUG: First 100 chars: {values_str[:100]}")
                    
                    # Parse each value group
                    for vg in value_groups:
                        row = self.parse_insert_values(vg)
                        if row:
                            rows.append(row)
                else:
                    print(f"Warning: Could not find end of INSERT statement {i+1} for {table_name}")
                            
            return rows
        except Exception as e:
            print(f"Erreur parsing table {table_name}: {e}")
            return []
    
    def parse_insert_values(self, values_str):
        """Parse une ligne de valeurs INSERT"""
        try:
            # Enlever les parenthèses extérieures
            content = values_str.strip()[1:-1]  # Remove ( and )
            
            # Handle empty content
            if not content:
                return []
            
            # Split by comma but respect quotes
            values = []
            current = ""
            in_quote = False
            quote_char = None
            
            for char in content:
                if not in_quote and char in ('"', "'"):
                    in_quote = True
                    quote_char = char
                elif in_quote and char == quote_char:
                    # Check if it's escaped
                    if len(current) > 0 and current[-1] == '\\':
                        current += char  # Keep the escaped quote
                    else:
                        in_quote = False
                        quote_char = None
                        current += char  # Close quote
                elif not in_quote and char == ',':
                    values.append(current.strip())
                    current = ""
                else:
                    current += char
            
            if current:
                values.append(current.strip())
            
            # Clean up each value
            parsed_values = []
            for value in values:
                # Remove surrounding quotes if present
                if value.startswith("'") and value.endswith("'"):
                    value = value[1:-1]
                elif value.startswith('"') and value.endswith('"'):
                    value = value[1:-1]
                
                # Handle NULL values
                if value.upper() == 'NULL':
                    parsed_values.append(None)
                else:
                    # Unescape common SQL escape sequences
                    unescaped = value.replace("\\'", "'").replace("\\\\", "\\")
                    parsed_values.append(unescaped)
            
            return parsed_values
        except Exception as e:
            print(f"Erreur parsing values: {e}")
            return []
    
    def extract_all_data(self):
        """Extrait toutes les données des tables WordPress"""
        print("Extraction des données WordPress...")
        
        # Extraire les données de chaque table avec le préfixe correct
        self.posts_data = self.parse_sql_insert('wnew23p_posts')
        print(f"Posts: {len(self.posts_data)} enregistrements")
        
        self.terms_data = self.parse_sql_insert('wnew23p_terms')
        print(f"Terms: {len(self.terms_data)} enregistrements")
        
        self.term_taxonomy_data = self.parse_sql_insert('wnew23p_term_taxonomy')
        print(f"Term Taxonomy: {len(self.term_taxonomy_data)} enregistrements")
        
        self.postmeta_data = self.parse_sql_insert('wnew23p_postmeta')
        print(f"Postmeta: {len(self.postmeta_data)} enregistrements")
        
        self.term_relationships_data = self.parse_sql_insert('wnew23p_term_relationships')
        print(f"Term Relationships: {len(self.term_relationships_data)} enregistrements")
    
    def get_post_meta(self, post_id, meta_key):
        """Récupère une valeur meta pour un post"""
        for meta in self.postmeta_data:
            if len(meta) >= 4 and meta[1] == str(post_id) and meta[2] == meta_key:
                return meta[3]
        return ""
    
    def get_post_metas(self, post_id):
        """Récupère toutes les metas pour un post"""
        metas = {}
        for meta in self.postmeta_data:
            if len(meta) >= 4 and meta[1] == str(post_id):
                metas[meta[2]] = meta[3]
        return metas
    
    def get_term_name(self, term_id):
        """Récupère le nom d'un terme"""
        for term in self.terms_data:
            if len(term) >= 2 and term[0] == str(term_id):
                return term[1]
        return ""
    
    def get_category_path(self, term_id):
        """Construit le chemin complet d'une catégorie"""
        path_parts = []
        current_id = term_id
        
        # Chercher la catégorie parente
        for tt in self.term_taxonomy_data:
            if len(tt) >= 4 and tt[1] == str(current_id) and tt[3] == 'category':
                parent_id = tt[4] if len(tt) > 4 else "0"
                if parent_id != "0" and parent_id != current_id:
                    path_parts.append(self.get_term_name(current_id))
                    # Récursivement chercher le parent (simplifié)
                    parent_name = self.get_term_name(parent_id)
                    if parent_name:
                        path_parts.append(parent_name)
                else:
                    path_parts.append(self.get_term_name(current_id))
                break
        
        return " > ".join(reversed(path_parts))
    
    def generate_categories_csv(self):
        """Génère le CSV des catégories"""
        print("Génération du CSV des catégories...")
        
        categories = []
        processed_terms = set()
        
        # Traiter les taxonomies de type catégorie
        for tt in self.term_taxonomy_data:
            if len(tt) >= 4 and tt[3] == 'category':
                term_id = tt[1]
                if term_id not in processed_terms:
                    processed_terms.add(term_id)
                    
                    category_name = self.get_term_name(term_id)
                    category_path = self.get_category_path(term_id)
                    
                    # Récupérer la description
                    description = ""
                    for term in self.terms_data:
                        if len(term) >= 4 and term[0] == term_id:
                            description = term[3] if len(term) > 3 else ""
                            break
                    
                    categories.append({
                        'CATEGORIES FR': category_path or category_name,
                        'CATEGORIES UK': category_name,
                        'CHAPO FR': description[:200] + "..." if len(description) > 200 else description,
                        'CHAPO UK': description[:200] + "..." if len(description) > 200 else description,
                        'DESCRIPTIF FR': description,
                        'DESCRIPTIF UK': description,
                        'PHOTO': ""
                    })
        
        # Écrire le CSV
        with open(os.path.join(self.output_dir, 'categories.csv'), 'w', encoding='utf-8', newline='') as csvfile:
            fieldnames = ['CATEGORIES FR', 'CATEGORIES UK', 'CHAPO FR', 'CHAPO UK', 'DESCRIPTIF FR', 'DESCRIPTIF UK', 'PHOTO']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(categories)
        
        print(f"Categories CSV généré: {len(categories)} catégories")
    
    def generate_products_csv(self):
        """Génère le CSV des produits"""
        print("Génération du CSV des produits...")
        
        products = []
        product_count = 0
        
        # Traiter les posts de type produit
        for post in self.posts_data:
            if len(post) >= 22:  # Nombre de colonnes dans la table posts
                post_id = post[0]  # ID
                post_type = post[20]  # post_type est à l'index 20
                
                # Clean up post_type by removing extra quotes
                if post_type:
                    post_type = post_type.strip("'\"")
                
                # Debug: afficher les types de posts trouvés
                if post_type:
                    print(f"Type de post trouvé: '{post_type}' (repr: {repr(post_type)})")
                
                if post_type == 'product':
                    product_count += 1
                    post_title = post[5] if len(post) > 5 else ""  # post_title est à l'index 5
                    post_content = post[4] if len(post) > 4 else ""  # post_content est à l'index 4
                    post_excerpt = post[6] if len(post) > 6 else ""  # post_excerpt est à l'index 6
                    
                    # Handle encoding issues in print
                    try:
                        print(f"Produit trouvé: {post_title} (ID: {post_id})")
                    except UnicodeEncodeError:
                        print(f"Produit trouvé: [titre avec caractères spéciaux] (ID: {post_id})")
                    
                    # Récupérer les metas du produit
                    metas = self.get_post_metas(post_id)
                    
                    # Récupérer le prix
                    price = metas.get('_price', '0')
                    regular_price = metas.get('_regular_price', price)
                    sale_price = metas.get('_sale_price', '')
                    
                    # Récupérer la référence SKU
                    sku = metas.get('_sku', f'WP_{post_id}')
                    
                    # Récupérer les catégories
                    categories = []
                    for rel in self.term_relationships_data:
                        if len(rel) >= 3 and rel[0] == post_id:
                            term_taxonomy_id = rel[1]
                            for tt in self.term_taxonomy_data:
                                if len(tt) >= 4 and tt[0] == term_taxonomy_id and tt[3] == 'category':
                                    category_name = self.get_term_name(tt[1])
                                    if category_name:
                                        categories.append(category_name)
                    
                    category_path = " > ".join(categories) if categories else "NON CLASSÉ"
                    
                    # Récupérer la marque si disponible (comme attribut)
                    brand = ""
                    # Chercher dans les attributs WooCommerce
                    for meta_key, meta_value in metas.items():
                        if 'attribute' in meta_key.lower() and ('brand' in meta_key.lower() or 'marque' in meta_key.lower()):
                            brand = meta_value
                            break
                    
                    # Récupérer l'image
                    image_url = ""
                    thumbnail_id = metas.get('_thumbnail_id', '')
                    if thumbnail_id:
                        # Chercher l'URL de l'image dans les posts
                        for img_post in self.posts_data:
                            if len(img_post) >= 22 and img_post[0] == thumbnail_id and img_post[20] == 'attachment':
                                img_metas = self.get_post_metas(thumbnail_id)
                                image_url = img_metas.get('_wp_attached_file', '')
                                break
                    
                    product_data = {
                        'REF': sku,
                        'TITRE UK': post_title,
                        'CHAPO UK': post_excerpt[:200] + "..." if len(post_excerpt) > 200 else post_excerpt,
                        'CHAPO FR': post_excerpt[:200] + "..." if len(post_excerpt) > 200 else post_excerpt,
                        'DESCRIPTIF UK': post_content,
                        'DESCRIPTIF FR': post_content,
                        'POSTSCRIPTUM UK': '',
                        'POSTSCRIPTUM FR': '',
                        'PRIX': regular_price if regular_price else price,
                        'PRIX2': sale_price if sale_price else '',
                        'PHOTO': image_url.split('/')[-1] if image_url else '',
                        'BRAND': brand,
                        'COULEUR UK': '',
                        'MATERIAL UK': '',
                        'CONTENT UK': '',
                        'CATEGORIE': category_path
                    }
                    
                    products.append(product_data)
        
        print(f"Nombre total de produits trouvés: {product_count}")
        
        # Écrire le CSV
        with open(os.path.join(self.output_dir, 'products.csv'), 'w', encoding='utf-8', newline='') as csvfile:
            fieldnames = ['REF', 'TITRE UK', 'CHAPO UK', 'CHAPO FR', 'DESCRIPTIF UK', 'DESCRIPTIF FR', 
                         'POSTSCRIPTUM UK', 'POSTSCRIPTUM FR', 'PRIX', 'PRIX2', 'PHOTO', 'BRAND', 
                         'COULEUR UK', 'MATERIAL UK', 'CONTENT UK', 'CATEGORIE']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(products)
        
        print(f"Products CSV généré: {len(products)} produits")
    
    def generate_brands_csv(self):
        """Génère le CSV des marques"""
        print("Génération du CSV des marques...")
        
        # Extraire les marques des attributs de produits
        brands = set()
        for post in self.posts_data:
            if len(post) >= 13 and post[12] == 'product':
                post_id = post[0]
                metas = self.get_post_metas(post_id)
                
                for meta_key, meta_value in metas.items():
                    if 'attribute' in meta_key.lower() and ('brand' in meta_key.lower() or 'marque' in meta_key.lower()):
                        if meta_value and meta_value != '':
                            brands.add(meta_value)
        
        # Créer le CSV des marques
        brands_data = [{'brand': brand} for brand in sorted(brands)]
        
        with open(os.path.join(self.output_dir, 'brands.csv'), 'w', encoding='utf-8', newline='') as csvfile:
            fieldnames = ['brand']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(brands_data)
        
        print(f"Brands CSV généré: {len(brands_data)} marques")
    
    def generate_colors_csv(self):
        """Génère le CSV des couleurs"""
        print("Génération du CSV des couleurs...")
        
        # Extraire les couleurs des attributs de produits
        colors = set()
        for post in self.posts_data:
            if len(post) >= 13 and post[12] == 'product':
                post_id = post[0]
                metas = self.get_post_metas(post_id)
                
                for meta_key, meta_value in metas.items():
                    if 'attribute' in meta_key.lower() and ('color' in meta_key.lower() or 'couleur' in meta_key.lower()):
                        if meta_value and meta_value != '':
                            colors.add(meta_value)
        
        # Créer le CSV des couleurs
        colors_data = [{'color': color} for color in sorted(colors)]
        
        with open(os.path.join(self.output_dir, 'colors.csv'), 'w', encoding='utf-8', newline='') as csvfile:
            fieldnames = ['color']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(colors_data)
        
        print(f"Colors CSV généré: {len(colors_data)} couleurs")
    
    def generate_materials_csv(self):
        """Génère le CSV des matériaux"""
        print("Génération du CSV des matériaux...")
        
        # Extraire les matériaux des attributs de produits
        materials = set()
        for post in self.posts_data:
            if len(post) >= 13 and post[12] == 'product':
                post_id = post[0]
                metas = self.get_post_metas(post_id)
                
                for meta_key, meta_value in metas.items():
                    if 'attribute' in meta_key.lower() and ('material' in meta_key.lower() or 'matiere' in meta_key.lower() or 'matériau' in meta_key.lower()):
                        if meta_value and meta_value != '':
                            materials.add(meta_value)
        
        # Créer le CSV des matériaux
        materials_data = [{'material': material} for material in sorted(materials)]
        
        with open(os.path.join(self.output_dir, 'materials.csv'), 'w', encoding='utf-8', newline='') as csvfile:
            fieldnames = ['material']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(materials_data)
        
        print(f"Materials CSV généré: {len(materials_data)} matériaux")
    
    def convert_all(self):
        """Lance la conversion complète"""
        print("Début de la conversion WordPress vers Thelia...")
        
        # Extraire toutes les données
        self.extract_all_data()
        
        # Générer tous les fichiers CSV
        self.generate_categories_csv()
        self.generate_products_csv()
        self.generate_brands_csv()
        self.generate_colors_csv()
        self.generate_materials_csv()
        
        print(f"Conversion terminée! Fichiers CSV générés dans {self.output_dir}/")

if __name__ == "__main__":
    converter = WordPressToTheliaConverter("setup/uvpxgnzzfh.sql")
    converter.convert_all()
